### PR TITLE
feat(lsp):  run "source.fixAll.ast-grep" onsave

### PR DESCRIPTION
**Pull request by SoonIter**
https://github.com/ast-grep/ast-grep-vscode/issues/67
format with ast-grep when setting source.fixAll
```jsonc
// .vscode/settings.json
{
  "editor.codeActionsOnSave": {
    "source.fixAll.ast-grep": "explicit" // or "source.fixAll" it works too
  }
}
```

## Summary by CodeRabbit

- **Refactor**
    - Improved code action generation logic for efficiency.
- **New Features**
    - Expanded code action options to include source-related actions.
- **Chores**
    - Modified `Cargo.toml` file for better dependency readability.

### CODE REVIEW

File: crates/lsp/src/lib.rs [hunk 1]
```diff
@@ -383,26 +399,66 @@ impl<L: LSPLang> Backend<L> {
...
+    let diagnostics = params.context.diagnostics;
+    let error_id_to_ranges = Self::build_error_id_to_ranges(diagnostics);
+    let mut response = CodeActionResponse::new();
+
+    let code_action = params.context.only.as_ref()?.first();
+
+    // we only handle these code_actions
+    // 1. QuickFix
+    // 2. "source.fixAll" and "source.fixAll.ast-grep"
+    match code_action {
```
Comment (HerringtonDarkholme) [inline crates/lsp/src/lib.rs:423]: NIT: we can use `if matches!(xxx)` to simplify this code

File: crates/lsp/src/lib.rs [hunk 2]
```diff
@@ -383,26 +399,66 @@ impl<L: LSPLang> Backend<L> {
...
+        {
+          return Some(response);
+        }
+      }
+      None => {
+        return Some(response);
       }
     }
+
+    let rules = if let Ok(rules) = &self.rules {
```
Comment (HerringtonDarkholme) [inline crates/lsp/src/lib.rs:437]: NIT:
```suggestion
    let Ok(rules) = &self.rules else {
      return Some(response);
    }
```

File: crates/lsp/src/lib.rs [hunk 3]
```diff
@@ -67,6 +68,15 @@ impl<L: LSPLang> Backend<L> {
 
 const FALLBACK_CODE_ACTION_PROVIDER: Option<CodeActionProviderCapability> =
   Some(CodeActionProviderCapability::Simple(true));
+
+trait AstGrepActionKind {
+  const SOURCE_FIX_ALL_AST_GREP: CodeActionKind;
+}
+
+impl AstGrepActionKind for CodeActionKind {
+  const SOURCE_FIX_ALL_AST_GREP: CodeActionKind = CodeActionKind::new("source.fixAll.ast-grep");
```
Comment (HerringtonDarkholme) [inline crates/lsp/src/lib.rs:77]: QQ: why do we need a new fixAll variant for ast-grep? Should it be implemented as fixAll?

File: crates/lsp/src/lib.rs [hunk 4]
```diff
@@ -67,6 +68,15 @@ impl<L: LSPLang> Backend<L> {
 
 const FALLBACK_CODE_ACTION_PROVIDER: Option<CodeActionProviderCapability> =
   Some(CodeActionProviderCapability::Simple(true));
+
+trait AstGrepActionKind {
```
[the remark this answers is not in the document]
Comment (SoonIter) [reply crates/lsp/src/lib.rs:72]: just for usage of `CodeActionKind::SOURCE_FIX_ALL_AST_GREP`, like `CodeActionKind::QUICKFIX` and `CodeActionKind::SOURCE_FIX_ALL`. Is there a better way?

File: crates/lsp/src/lib.rs (hunk 3 above)
Comment (SoonIter) [reply crates/lsp/src/lib.rs:77]:
```json
 "editor.codeActionsOnSave": {
    "source.fixAll.eslint": "explicit"
  },
```
I think this is a specification of vscode.
https://github.com/microsoft/vscode-eslint/blob/c7ab1ded33213a43ce79b5caa4372ffa4ac98dc3/server/src/eslintServer.ts#L590-L591
https://github.com/ast-grep/ast-grep-vscode/issues/67#issuecomment-1943096215

In this way, they can also be defined in order of execution.
```json
 "editor.codeActionsOnSave": [
 	 "source.organizeImports", "source.fixAll.eslint",  "source.fixAll.ast-grep"
 ]
```

File: crates/lsp/src/lib.rs (hunk 1 above)
Comment (SoonIter) [reply crates/lsp/src/lib.rs:423]: I questioned chatGPT, and he told me it should be simplified as below. 😂how do you think?
```rust
	let code_actions = params.context.only.as_deref();
    if !matches!(
      code_actions,
      Some(&[
        CodeActionKind::QUICKFIX
          | CodeActionKind::SOURCE_FIX_ALL
          | CodeActionKind::SOURCE_FIX_ALL_AST_GREP,
        ..
      ])
    ) {
      return Some(response);
    }
```

File: crates/lsp/src/lib.rs [hunk 5]
```diff
@@ -383,26 +399,66 @@ impl<L: LSPLang> Backend<L> {
...
-          diagnostics: None,
-          edit,
-          disabled: None,
-          kind: Some(CodeActionKind::QUICKFIX),
-          is_preferred: Some(true),
-          data: None,
-        };
-        response.push(CodeActionOrCommand::from(action));
+        changes
+          .entry(text_document.uri.clone())
```
[the remark this answers is not in the document]
Comment (SoonIter) [reply crates/lsp/src/lib.rs]: done

File: crates/lsp/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SoonIter) [reply crates/lsp/src/lib.rs]: I only keep this line
```rust
const SOURCE_FIX_ALL_AST_GREP: CodeActionKind = CodeActionKind::new("source.fixAll.ast-grep");
```

File: crates/lsp/src/lib.rs (hunk 1 above)
Comment (SoonIter) [reply crates/lsp/src/lib.rs]: I modified it
```rust
  if code_action != CodeActionKind::QUICKFIX
      && code_action != CodeActionKind::SOURCE_FIX_ALL
      && code_action != SOURCE_FIX_ALL_AST_GREP
    {
      return Some(response);
    }
```

**Review by HerringtonDarkholme — APPROVED** [review]

**Pull request merged**